src/dune_engine/scheduler.ml

```diff
@@ -643,8 +676,101 @@ module Handler = struct
   type t = Config.t -> Event.t -> unit
 end
 
+module Timer : sig
+  type t
+
+  val create : Event.Queue.t -> frequency:float -> t
+
+  type sleeper
+
+  val await : sleeper -> [ `Finished | `Cancelled ] Fiber.t
+
+  val cancel : t -> sleeper -> unit Fiber.t
+
+  val sleep : t -> float -> sleeper
+
+  val close : t -> unit
+end = struct
+  type sleeper = [ `Finished | `Cancelled ] Fiber.Ivar.t
+
+  type t =
+    { events : Event.Queue.t
+    ; mutex : Mutex.t
+    ; frequency : float
+    ; mutable sleepers : (float * [ `Finished | `Cancelled ] Fiber.Ivar.t) list
+    ; mutable active : bool
+    }
+
+  let await = Fiber.Ivar.read
+
+  let cancel t sleeper =
+    Mutex.lock t.mutex;
+    t.sleepers <-
+      List.filter t.sleepers ~f:(fun (_, sleeper') -> not (sleeper' == sleeper));
+    Mutex.unlock t.mutex;
+    let* res = Fiber.Ivar.peek sleeper in
+    match res with
+    | None -> Fiber.Ivar.fill sleeper `Cancelled
+    | Some _ -> Fiber.return ()
+
+  let polling_loop t () =
+    let rec loop () =
+      Mutex.lock t.mutex;
+      match t.active with
+      | false -> ()
+      | true ->
+        let now = Unix.gettimeofday () in
+        t.sleepers <-
+          List.filter_map t.sleepers ~f:(fun (expiration, ivar) ->
+              if now < expiration then
+                Some (expiration, ivar)
+              else (
+                Event.Queue.send_timer_completed t.events
+                  (Fiber.Fill (ivar, `Finished));
+                None
+              ));
+        Mutex.unlock t.mutex;
+        Thread.delay t.frequency;
+        loop ()
+    in
+    loop ();
+    List.iter t.sleepers ~f:(fun (_, ivar) ->
+        Event.Queue.send_timer_completed t.events
+          (Fiber.Fill (ivar, `Cancelled)));
+    t.sleepers <- [];
+    Mutex.unlock t.mutex
+
+  let create events ~frequency =
+    let t =
+      { events
+      ; active = true
+      ; sleepers = []
+      ; frequency
+      ; mutex = Mutex.create ()
+      }
+    in
+    Thread.spawn (polling_loop t);
+    t
+
+  let sleep t duration =
+    Mutex.lock t.mutex;
+    let ivar = Fiber.Ivar.create () in
+    if not t.active then
+      Code_error.raise "cannot schedule timers after close" [];
```

Comment:
Need to unlock the mutex here.
In fact maybe it's worth scanning the rest of the code for mutex release in case of exception.